Respect the refactor of Android Streaming SDK (#18)
anbox-cloud/discourse-docs: client-side-virtual-keyboard.md

## Integrate components

> After importing the AAR file, add the `AnboxWebView` element to the activity layout file and adjust its layout ot meet your needs. For example:

Author: Thanks. fixed.

        android:layout_height="match_parent" />
   ```

## Build communication bridge

> In your HTML, define the JavaScript functions that can be invoked from the Android Java layer (AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. This builds up a communication tunnel between Android Java layer (AnboxWebView) and JavaScript layer.

Author: Fixed.

> In your HTML code, define the JavaScript functions that can be invoked from the Android Java layer (AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. Defining them in your code builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer.

Reviewer: ```suggestion
In your web application code, define the JavaScript functions that can be invoked from the Android Java layer (AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK. Defining them in your code builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer.
```

Author: Will fix.

   ```
   <script>
     function sendIMECommittedText(text) {

>        window.stream.sendIMECommittedText(text);

Reviewer: We'll want to specify where that `window.stream` comes from

     }

>      function sendIMEComposingText(text) {

Reviewer: This is a bit confusing. As reader, I have a few questions:
1. What are those functions ? You mention `define the JavaScript functions that can be invoked from the Android Java layer` but it's not specified whether the user has to copy paste the given code snippet, or if he has to implement something.
2. Where should these functions be defined ? What/who calls those functions ? Some more complex web apps might not want to expose those functions in a `<script></script>` tag

Author: What are those functions ? You mention define the JavaScript functions that can be invoked from the Android Java layer but it's not specified whether the user has to copy paste the given code snippet, or if he has to implement something.
A: User doesn't need to implement actual logic how IME text is sent from Android java layer to JavaScript, as the caller, they just need to define those functions, which acts as a bridge for the communication between Android application(more specifically com.android.) and JS SDK through the HTML code.
2. Where should these functions be defined ? 
A: From the context, those functions must be defined in your HTML(JS) code, 
4. What/who calls those functions ? 
A: AnboxWebView component which is provided from `com.canonical.anbox.streaming_sdk.aar` is going to call those functions. That means for most common case, in the actual java code, developers just need to import the `com.canonical.anbox.streaming_sdk.aar` file and use AnboxWebView in their layout XML file. That's it. And they don't need to care about the protocol that is used on both ends. We will help them to figure out that, that's the major change(refractor) we made in 1.11.2. I'd suggest taking a look at this change and commit message for the reason behind the sense.
https://code.launchpad.net/~gary-wzl77/indore/+git/anbox-streaming-stack/+merge/407715
6. Some more complex web apps might not want to expose those functions in a <script></script> tag
A: Web developers don't need to implement those functions if they don't want to integrate client-side virtual keyboard. That's being said, if you want to integrate the client side virtual keyboard, following the interface definition, provide those functions in your HTML(JS) code, everything should be working under the hood. If you don't want to expose those functions, there is no way for the Android java layer to talk to the js sdk to communicate with Android container through the tunnel

Reviewer: **they just need to define those functions**
I see, so it's just a matter of wording. Instead of 
`define the JavaScript functions that can be invoked from the Android Java layer`
we could have something like 
`copy the following Javascript functions in your web application. They will automatically be invoked by the Android java layer.`

**provide those functions in your HTML(JS) code**
That's the thing I was getting at, we should specify that those functions must be attached to the global `window` object. For example, if a user defines those function elsewhere, e.g. `window.someFramework.views.main.anboxIntegration` then this would not work

Author: how about
```
Adding the following JavaScript code snippet builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer. It attaches the AnboxStream object to the global `window` variable. And those functions will automatically be invoked by the Android java layer(AnboxWebView).
```

       window.stream.sendIMEComposingText(text);
     }
     function sendIMETextDeletion(counts) {
       window.stream.sendIMETextDeletion(counts);
     }
     function sendIMEAction(name, params) {
       window.stream.sendIMEAction(name, params);
     }
     function sendIMEComposingRegion(start, end) {
       window.stream.sendIMEComposingRegion(start, end);
     }
   </script>

   ```

After you build your project, run the application and stream from Anbox Cloud,
* when the text edit of the application in the container gains the focus, the client-side virtual keyboard pops up accordingly, and vice versa.

> * when typing with client-side virtual keyboard, the input text will be sent to the Android container and displayed in the running application.

Reviewer: ```suggestion
* when typing on the client-side virtual keyboard, the input text will be sent to the Android container and displayed in the running application.
```

> The above workflow provides a general overview of how to integrate the client-side virtual keyboard feature into an Android application. For the complete implementation details, refer to the `enhanced_webview_streaming` example in the `example/android` folder.

Reviewer: ```suggestion
The above workflow provides a general overview on how to integrate the client-side virtual keyboard feature into an Android application. For details on the complete implementation, refer to the `enhanced_webview_streaming` example in the `example/android` folder.
```

Author: Will fix

> Furthermore, by taking advantage of the interfaces that the AnboxWebView exposes, you can easily provide your own implementation of

Reviewer: This should be in a separate section, something like `## Advanced` or `## Customizing the virtual keyboard`

Author: +1.
Will fix

        android:layout_height="match_parent" />
   ```

## 3. Build communication bridge

In your web application code, define the JavaScript functions that can be invoked from the Android Java layer (AnboxWebView). Those functions are just wrappers of the APIs that are exposed from the JavaScript SDK.

Adding the following JavaScript code snippet builds up a communication tunnel between the Android Java layer (AnboxWebView) and the JavaScript layer. It attaches the AnboxStream object to the global `window` object. The functions will automatically be invoked by the Android Java layer (AnboxWebView).

   ```
   <script type="module">
     import {AnboxStream} from './js/anbox-stream-sdk.js';
     window.stream = new AnboxStream({
     ...
     ...
     ...
     });
   </script>

   <script>

>      function sendIMECommittedText(text) {

Reviewer: Nitpick: it would be nice to say where those function should be defined, e.g.:

> If you define those functions somewhere else, make sure they are stored under the global `window` object, e.g.: running `window.sendIMECommittedText()` should work

Author: That's fine.  However as we've shown how to initialize the stream object and attach the stream object to the global window object, I'd prefer to keep it as is and not go further down to the actual JS related knowledge since the section is mainly about the client-side virtual keyboard integration.

Reviewer: Fair enough, LGTM then
